### ISSUE TITLE
Building simple TS file randomly throws 3221226356 exit code

### ISSUE THREAD

**Issue by crutchcorn** [parcel-bundler/parcel#7578]
# 🐛 bug report

When building a relatively simple TypeScript file, it will build some of the time, but not others. For example, given the following TS file:

```typescript
const config = {
  serverPort: 3000,
};

export default config;
```

It will semi-randomly throw the error code 3221226356

## 🎛 Configuration (.babelrc, package.json, cli command)

```json
{
  "name": "@streampods/internalconfig",
  "description": "For internal use only. To share simple values between internal apps",
  "version": "0.0.0",
  "private": true,
  "source": "index.ts",
  "main": "dist/main.js",
  "module": "dist/module.js",
  "types": "dist/types.d.ts",
  "scripts": {
    "build": "parcel build"
  },
  "devDependencies": {
    "@parcel/packager-ts": "^2.2.1",
    "@parcel/transformer-typescript-types": "^2.2.1",
    "parcel": "^2.2.1",
    "typescript": "^4.5.4"
  }
}
```

## 🤔 Expected Behavior

No error should be thrown

## 😯 Current Behavior

When running build repeatedly, I get different results, but most of the time results in an exit code

```
 🌈    git  repro  config  main ≢ +6 | ~1 -1 
↳  yarn build
yarn run v1.22.17
$ parcel build
√ Built in 480ms

dist\main.js       557 B    142ms
dist\module.js     239 B    113ms
dist\types.d.ts    115 B     24ms
Done in 1.01s.
 🌈    git  repro  config  main ≢ +6 | ~1 -1 
↳  yarn build
yarn run v1.22.17
$ parcel build
√ Built in 469ms
error Command failed with exit code 3221226356.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
 🌈    git  repro  config  main ≢ +6 | ~1 -1   
↳  yarn build
yarn run v1.22.17
$ parcel build
√ Built in 464ms

dist\main.js       557 B    142ms
dist\module.js     239 B    113ms
dist\types.d.ts    115 B     24ms
error Command failed with exit code 3221226356.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

## 💁 Possible Solution

Unfortunately, I have no idea what might be causing this problem.

## 🔦 Context

I'm trying to use `parcel` to build a internal dependency in a TurboRepo monorepo for usage in both NodeJS (CJS) and Browser (ESM) applications

## 💻 Code Sample

Minimal reproduction lives here:

https://github.com/oceanbit/parcel-3221226356-repro

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | `2.2.1`
| Node             | `16.13.1`
| npm/Yarn         | `1.22.17`
| Operating System | Windows 11

**Comment by jakubmazanec** [comment]
Same problem. Sometimes the exit code yarn reports is `3221225477`.

**Comment by astegmaier** [comment]
I get the same problem (with the `3221226356` exit code in the original bug. It doesn't happen 100% of the time - maybe more like 50%+ - and I haven't noticed a pattern yet. Just like the original post, I have a relatively small number of typescript files in my project. Here's my environment:

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.2.1
| Node             | 14.18.2
| Yarn         | 1.22.17
| Operating System | Win11

**Comment by Jurasinho** [comment]
Same problem. 

Software | Version(s)
-- | --
Parcel | 2.2.1
Node | 14.8.0
NPM | 6.14.7
Operating System | Win10

**Comment by olhapi** [comment]
same issue

<html>
<body>


Software | Version(s)
-- | --
Parcel | 2.2.1
Node | v16.13.0
Yarn | 1.22.15
Operating System | Win11


</body>
</html>

**Comment by marco-verbeek** [comment]
Same issue here. 
Temp solution: run `yarn build` until it eventually works. (fails randomly approx. 9x on 10 tries)

<html>
<body>


Software | Version(s)
-- | --
Parcel | 2.2.1
Node | v16.13.1
Yarn | 1.22.17
Operating System | Win10


</body>
</html>

**Comment by kilianplapp** [comment]
Same issue
Solution, as @marco-verbeek said, is to keep running it until it works.

**Comment by Systemcluster** [comment]
Parcel version `2.0.0-rc.0` does not show this issue according to some limited testing.

**Comment by devongovett** [comment]
I've been trying to reproduce but haven't had any success so far. The repro linked in the original issue builds 100% of the time for me on Windows 10, node 16.13.2. Anyone have any other details about your setups that might be relevant?

**Comment by davidwincent** [comment]
I have the same problem with a minimal swagger-ui project.
[parcel-7578.zip](https://github.com/parcel-bundler/parcel/files/7931640/parcel-7578.zip)

yarn start fails 9 of 10 times on Windows. I do not see this problem with latest parcel in Docker/Linux environment.

After downgrading parcel to 2.0.0-rc.0 yarn start fails 0 of 10 times.



| Software | Version |
| ------------- | ------------- |
| parcel  | 2.2.1  |
| NodeJS  | 16.13.0  |
| npm  | 8.1.0  |
| yarn  | 1.22.17  |
| Windows | 10.20H2.19042.1466  |

**Comment by kai-dorschner-twinsity** [comment]
Hey @devongovett 
> I've been trying to reproduce but haven't had any success so far. The repro linked in the original issue builds 100% of the time for me on Windows 10, node 16.13.2. Anyone have any other details about your setups that might be relevant?

Unfortunately I can't give you our company source code where this happens and in a test repo I can't reproduce it RN, but I'll keep syncing the parameters and let you know.

FYI, stackoverflow says:
> 3221225477 converted to hex is 0xC0000005, which stands for STATUS_ACCESS_VIOLATION, which means you tried to access (read, write or execute) invalid memory.
https://stackoverflow.com/a/53369553

Edit:
The error code `3221226356` translates to hex `0xC0000374` and means heap corruption.

**Comment by kai-dorschner-twinsity** [comment]
Update @devongovett OK I got an example project working https://github.com/lookmom/nohands
It ONLY worked after I did `npm upgrade` on the root. Which means some subdependency got updated that sometimes produces the `3221225477` (and sometimes `3221226356`) on `npm run build`. Note that it happens completely randomly. Nothing helped to mitigate, reduce or provoke the error.

You'll need Win10 and I used Node `v17.4.0` and npm `8.3.2`

**Comment by devongovett** [comment]
This example also builds fine for me on Windows 10 20H2...

**Comment by davidwincent** [comment]
@devongovett have you tried the source code I uploaded in parcel-7578.zip ?

**Comment by kai-dorschner-twinsity** [comment]
Can you try to also do an `npm upgrade` in the project root? (though it shouldn't change anything anymore).
Hint: I got Win 10 21H2

**Comment by devongovett** [comment]
@davidwincent your example also works for me... 😕

@kai-dorschner-twinsity your example continues to work after npm upgrade.

**Comment by kai-dorschner-twinsity** [comment]
What else could it be? I am using Windows Terminal (1.11.3471.0) instead of plain cmd. Is the MS Visual C++ version relevant (14.29.3019)? Python (none installed)?

**Comment by davidwincent** [comment]
Have tried the following:

1. remove yarn.lock and yarn install
2. remove yarn.lock and npm install && npm upgrade

with npm start I get a silent exit
with yarn start I get the following:
'''
>yarn start
yarn run v1.22.17
$ parcel serve -p 8082 --dist-dir serve ./index.html
error Command failed with exit code 3221225477.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
'''

**Comment by davidwincent** [comment]
BTW. I am running as a non-priveleged user account.

3221225477 could be access violation on Windows?
https://github.com/nodejs/node/blob/842e5a10ac59c826f875ef79e92d144e2655cc9a/test/common.js#L502-L505

**Comment by kai-dorschner-twinsity** [comment]
After nuking the node_modules it starts back working. I prepared some steps to reproduce the error on Win 10 21H2:
1. `git clone https://github.com/lookmom/nohands.git`
2. `npm install`
3. `npm run build` # works this time
4. `npx rimraf ./**/node_modules` # nuke node_modules
5. `git reset d352731 --hard` # reset to previous commit
6. `npm install`
7. `git reset 9eb3f67 --hard` # reset back to the origin/HEAD
8. `npm install`
9. `npm run build` # fails this time

The error will be `ERROR: "build:with-scope-hoist" exited with 3221225477.`

I have a 100% Failure Rate with these steps

**Comment by davidwincent** [comment]
I am able to reproduce `error Command failed with exit code 3221226356.` on a second machine running Win11 as an Administrator user

The only fix I have found so far is to downgrade parcel using
`yarn add -D parcel@2.0.0-rc.0`

| Software | Version |
| --- | --- |
| parcel | 2.2.1 |
| NodeJs | 17.4.0 |
| yarn | 1.22.17 |
| npm | 8.3.1 |
| Windows | 11.21H2.22000.434 |

Edit exit code 3221226356 and 3221225477 is clearly not the same error.

On the Win10 machine I get exit code 3221225477 
On the Win11 machine I get exit code 3221226356

Both problems are solved by downgrading parcel.

**Comment by devongovett** [comment]
Unfortunately I still cannot reproduce with those steps on my machine. I am running Windows 10 20H2. Maybe something in 21H2? I don't have access to that update yet. Can you confirm which exact windows version you are on?

**Comment by devongovett** [comment]
Does it work for anyone if you set the `PARCEL_WORKER_BACKEND=process` environment variable?

**Comment by olhapi** [comment]
> Does it work for anyone if you set the PARCEL_WORKER_BACKEND=process environment variable?

I can confirm that it works for me with this env variable.

My setup: https://github.com/parcel-bundler/parcel/issues/7578#issuecomment-1019279955

**Comment by davidwincent** [comment]
@devongovett setting environment variable `PARCEL_WORKER_BACKEND=process` works for me!

Edit: confirmed on Win10.20H2.19042.1466 and Win11.21H2.22000.434

**Comment by kai-dorschner-twinsity** [comment]
> Does it work for anyone if you set the `PARCEL_WORKER_BACKEND=process` environment variable?

Can confirm it works, too. What does that mean for the further progress of this ticket? Should we keep this variable always on (even in CI and non-windows dev systems)? Do you know in which module that occurs and maybe can now force to reproduce that error to then fix whatever goes berserk? Can we provide useful information?

**Comment by kai-dorschner-twinsity** [comment]
> Unfortunately I still cannot reproduce with those steps on my machine. I am running Windows 10 20H2. Maybe something in 21H2? I don't have access to that update yet. Can you confirm which exact windows version you are on?

![image](https://user-images.githubusercontent.com/92744645/151163367-9aa62b4f-8ce1-4bb7-8433-8a0e086aa798.png)

Just plain old windows. No insider preview, no fancy stuff. The machine was installed completely new 20 days ago, so not even any migrations happened from older versions

**Comment by devongovett** [comment]
> What does that mean for the further progress of this ticket?

Well, at least we have a workaround now and some lead about the potential cause. It definitely seems to be a native dependency that is not thread safe. The question now is which one, so the next step would be to narrow that down. However, I still cannot reproduce locally, so that makes this slightly challenging...

**Comment by devongovett** [comment]
I think we narrowed this down to an upgrade of napi-rs in the `@parcel/source-map` package, which was released sometime after Parcel 2.2.1. I still don't know the exact root cause, but for now I've reverted this upgrade and published `@parcel/source-map` v2.0.2. See https://github.com/parcel-bundler/source-map/pull/91.

You'll need to remove `@parcel/source-map` 2.0.1 from your lock file and re-install to get the upgrade. Please comment whether this works for you. Apologies again for the issue, and thanks for your patience while we track it down.

**Comment by kilianplapp** [comment]
Just did a few quick tests, and it appears to have fixed the issue.
